### DISCUSSION TITLE
xmake install before xpack packaging

### DISCUSSION THREAD

**Discussion by jfulem** [xmake-io/xmake#5698] [Q&A]
Please, is it possible to somehow call `xmake call` instead of `xmake build`. When making the package (NSIS installer) we need to know all dependencies (*.dll) which are copied during the installation.
I tried to use:

```cpp
 on_buildcmd(function (package, batchcmds)
     batchcmds:run("xmake", {"config", "-m", "release", "--shared=yes", "-y"})
    batchcmds:run("xmake", {"build", "-y"})
    batchcmds:run("xmake", {"install", "-o", "installation"})
    )

add_installfiles("installation/bin/*.*", {prefixdir = "bin"})
add_installfiles("installation/lib/*.*", {prefixdir = "lib"})
```
But this causes the process to be stucked.

**Comment by waruqi** [comment]
you need not call them. xpack will install all dll deps in targets if you use `add_targets`.

https://github.com/xmake-io/xmake/blob/dev/tests/plugins/pack/xmake.lua#L29

**Reply by jfulem** [reply]
Well, that's the case. When I use `add_targets`  it installs only the `deps`. It does not install the dlls from the `packages`.
```
target("marker")
    -- buildinh executable
    set_kind("binary")
    -- add required packages
    add_packages("zeromq", "cppzmq", "flatbuffers", "nlohmann_json", "protobuf-cpp", "opencv") 
```

When I call: `xmake install` it does install also many opencv dlls, libprotobuf and zeromq dll. But when I use the `add_targets` in xpack I see only the libraries and executables I specified as targets in xmake.lua, there are no libraries from `add_packages`.

**Reply by waruqi** [reply]
you can debug https://github.com/xmake-io/xmake/blob/687beddbf316b1565a55b68934c3681320a1d51f/xmake/plugins/pack/batchcmds.lua#L202

**Comment by jfulem** [comment]
@waruqi, it tried to debug it. It looks the problem is on this line:
```
table.remove_if(libfiles, function (_, libfile) return not depends:has(path.filename(libfile)) end)
```
[here](https://github.com/xmake-io/xmake/blob/fcc2809210622102cdc0ec64256ff656b51eb509/xmake/plugins/pack/batchcmds.lua#L105)

All relevant files are removed.

**Reply by waruqi** [reply]
these dlls are unused for target. so xmake will strip them.

https://github.com/xmake-io/xmake/blob/fcc2809210622102cdc0ec64256ff656b51eb509/xmake/plugins/pack/batchcmds.lua#L101

you can use `set_policy("install.strip_packagelibs", false)` to disable it.

**Comment by jfulem** [comment]
@waruqi, thanks. When I use your code of `batchcmds.lua` from fcc2809 everything works fine. I can even use `set_policy("install.strip_packagelibs", true)`

**Comment by jfulem** [comment]
Hmm, it looks sometimes it works and sometimes it does not. The same for the `xmake install`. When I just delete the `build` directory and rerun the `build & install` it completely avoids the dependent `.dll` libraries.

**Comment by jfulem** [comment]
When I run `xmake install` and it runs OK, it installs all these libraries:
![image](https://github.com/user-attachments/assets/dbcc2675-ec81-4d02-9c7c-a9bd12273056)
when I run `xmake pack` it installs only:
`libzmq-v143-mt-4_3_5.dll`

When I debug it it looks in this part:
```
-- we can only reserve used libraries
    if project.policy("install.strip_packagelibs") then
        if target:is_binary() or target:is_shared() or opt.binaryfile then
            local depends = hashset.new()
            _get_target_package_deplibs(opt.binaryfile or target:targetfile(), depends, libfiles, {plat = target:plat(), arch = target:arch()})
            table.remove_if(libfiles, function (_, libfile) return not depends:has(path.filename(libfile)) end)
        end
    end
```
the `libfiles` and `depends` have no common libraries. But I don't understand how this is possible.

**Reply by waruqi** [reply]
you can debug here 

https://github.com/xmake-io/xmake/blob/86627f20f80a50fece956b4e8394afc4b5a83cd1/xmake/plugins/pack/batchcmds.lua#L67

https://github.com/xmake-io/xmake/blob/86627f20f80a50fece956b4e8394afc4b5a83cd1/xmake/modules/utils/binary/deplibs.lua#L26

**Reply by jfulem** [reply]
I tried, I spent really some time, but I'm not able to find the issue.

**Reply by waruqi** [reply]
Please provide a minimal reproducible project.
